src/librustc_metadata/dynamic_lib.rs

```diff
@@ -77,16 +77,35 @@ mod dl {
             // dlerror isn't thread safe, so we need to lock around this entire
             // sequence
             let _guard = (*LOCK).lock();
+
+            // dlerror reports the most recent failure that occured during a
+            // dynamic linking operation and then clears that error; we call
+            // once in advance of our operation in an attempt to discard any
+            // stale prior error report that may exist:
             let _old_error = libc::dlerror();
 
             let result = f();
 
-            let last_error = libc::dlerror() as *const _;
-            if ptr::null() == last_error {
+            // We should only check dlerror() in the event that the operation
+            // fails, which we determine by checking for a NULL return.  This
+            // covers at least dlopen() and dlsym().
```

Comment:
Can you document these semantics at the function level? Specifically, if `f` returns a null pointer, this function returns `Err` with the string in `dlerror`.

Also, just to be sure, do all the functions we pass to this helper return `NULL` and only `NULL` to indicate an error? There's no `(void *) 1` weirdness or something?

Reply:
For `dlsym` at least, the current approach is [explicitly recommended on linux](https://linux.die.net/man/3/dlsym) and seems to be necessary on illumos as well, since `NULL` can indicate either an error condition or a "symbol not found" error. We should be checking the return value of `dlopen`, but we will need to find a workaround here.